README
                                                      -*- indented-text -*-
Project:    ATmegaBOOT
----------------------------------------------------------------------------

This is an adaption of the bootloader code from http://chip45.com/. I've
changed the code to make it easier to configure and to match my requirements.

This code is a first try and I don't know if it is the last one. ;-)

To add support for a new product, copy the product_TEMPLATE.mak to a new file
and replace "TEMPLATE" with a unique shortcut for the products name. Change
the file to fit your need. If necessary, copy config_TEMPLATE.h and
isp_TEMPLATE.mak and replace "TEMPLATE" with the same shortcut.

Attention: be very carefull while editing the fuses in product_TEMPLATE.mak!
It is possible to damage your MCU!

To build the bootloader, call "make PRODUCT=TEMPLATE" and replace "TEMPLATE"
with the shortcut again. As a result, you will get three binaries (bin,
Intel-Hex and Motorola-SRecord).

Note: to configure the ISP settings, just edit config_isp.mak!

So the first steps for a product "FOO" could look like this:

   $ cp product_TEMPLATE.mak projects/product_FOO.mak
   $ editor product_FOO.mak
   $ cp isp_TEMPLATE.mak projects/isp_FOO.mak
   $ editor isp_FOO.mak
   $ cp config_TEMPLATE.h projects/config_FOO.h
   $ editor config_FOO.h
   $ make PRODUCT=FOO
   $ ls -1 ATmegaBOOT_FOO*
   $ ATmegaBOOT_FOO_V1_0.bin
   $ ATmegaBOOT_FOO_V1_0.hex
   $ ATmegaBOOT_FOO_V1_0.srec


Have fun.

--
jd@voelker-web.de


History
                                                      -*- indented-text -*-
+=========================================================================+
|                          DEVELOPMENT HISTORY                            |
+=========================================================================+

                  Project: ATmegaBOOT
                   Author: (changes by) J. Desch <github@jdesch.de>
          Current Release: 1.7

+=========================================================================+
|                         Copyright (c) J. Desch                          |
+=========================================================================+

2016/06/28 - jd

 * increased version to 1.7
 * add support for the Mega2560

2016/06/07 - jd

 * increased version to 1.6
 * FIX: while building the bootloader for a project running a Mega128, an error
   pops up. The Mega32/128/162 use a different name for MCUSR register (MCUCSR).
 * the bootloader and the "forced mode" is now tested for Mega2561 and Mega128.

2016/06/06 - jd

 * increased version to 1.5
 * add new "forced mode" to enter the bootloader. In the case there is no
   bootloader pin, the bootloader can be configured (USE_FORCED_BOOTLOAD_ENTER)
   to wait 3 seconds for a sequence of 5+ consecutive '*' at the UART.
   If this '*' sequence is received, the bootloader starts sending '*' and wait
   for an empty UART buffer. If no more data is fetched, the regular bootloader
   is entered.
 * add support for the Mega2561
 * add define BOOTLOADER_ABOVE_64K to MCU definitions

2015/04/14 - jd

 * add the optional configuration entry for the baudrate to the template
   of the product configuration.

2015/03/27 - jd

 * increased version to 1.4

 * fix: avr-gcc 4.7.2 complains about the wrong pointer size of PGM_P! Switched
   to uint_farptr_t for the Mega-128. Since the pointer are always 16bit, it
   isn't possible to get the correct address by simply casting the pointer.
   I've found the macro FAR(), which restaurates a 17bit address. The drawback
   of the macro is, that is only works with variables. It is not possible to
   pass a PSTR() constant to it. So I've put all messages into variables.


---- before hosted at github ----

2012/11/22 - jd		REL_1_3

 * fix: multiple define in the preprocessor are changed from #ifdef
   to #if. May be cleaner...

 * fix: some register names has changed in avr-libc 1.7.1

2012/05/10 - jd		REL_1_2

 * fix: to access UCSRC, the highest bit must be set! This is a register
   selection bit.

 * new configuration option WANT_AVRDUDE_SIGNATURE! Defining this enables
   the UNIVERSAL command. This is used by AVRDUDE to read out the
   signature of the device. The sequence is tested only with an ATmega32!

2012/05/03 - jd		REL_1_1B

 * fix in putsP for MCUs which doesn't support pgm_read_byte_far().
 * fix: getch() for single UART MCUs has used wrong registers.

2011/12/22 - jd		REL_1_1

 * fixed the putsP function (which didn't handle PGM_P strings)
 * welcome message is now global and stored in PROGMEM
 * getch now checks for "frame errors" and "overruns".
 * monitor's "echo" ('u') can be abort by pressing ESC.

2011/12/2 - jd		REL_1_0

 * due to the fact, that the upstream release isn't available anymore, we
   can ignore the ugly release id. So I've changed the current release
   "20111202" to V1.0.

2011/11/24 - jd

 * added support for m664, m644p and m324p with support for both uarts.

 * changed the SW_MINOR release of the firmware from 0x0F to 0x10.

 * if BL_PIN isn't defined, the pin isn't used. This means, the bootloader is
   allways entered. This is done to support starting the bootloader only from
   within the app.

2008/7/31 - jd

 * added configurable "quit behaviour"

 * minor changes to the makefiles

2008/7/29 - jd

 * made the codebase more configurable.

 * add support for RS485 transceiver Rx/Tx switching


+=========================================================================+
| END OF FILE                                                             |
+=========================================================================+


projects/config_WAAGE.h
/* config_WAAGE.h -- product specific bootloader configuration header.
 *
 * This header contains declarations and defines for a special product. To
 * enable the usage of this header, uncomment USE_CONFIG_H in the
 * product_TEMPLATE.mak!
 */


/* -------------------------------------------------------------------------
 * Some macros used by this header..
 */

/* This macro converts a sequence of 0 and 1 into a BYTE. The sequence start
 * with the MSB! Make shure that you allways define all 8 bits!
 * Sample: BIN2BYTE(11001010) is converted to 0xCA (1100->C & 1010->A).
 */
#define BIN2BYTE(a) ( ((0x##a##L>>21) & 0x80) + ((0x##a##L>>18) & 0x40) \
                    + ((0x##a##L>>15) & 0x20) + ((0x##a##L>>12) & 0x10) \
                    + ((0x##a##L>>9) & 0x08) + ((0x##a##L>>6) & 0x04)  \
                    + ((0x##a##L>>3) & 0x02) + (0x##a##L & 0x01))



/* -------------------------------------------------------------------------
 * CONFIGURATION
 */



/* ===== Configure Monitor Welcome =====
 * Uncomment to define the welcome message of the Monitor.
 */
/* #define WELCOME_MSG "MegaBOOT Monitor\r\n" */



/* ===== Configure the RS485 stuff =====
 * If your UART is a RS485, define the port-, the DDR-register and the pin number
 * of the output which enables (with 1) the RS485 transmitter. If you don't use a
 * RS485, comment out these three lines.
 */
/* #define RS485_TXON  PINE2 */
/* #define RS485_DDR   DDRE */
/* #define RS485_PORT  PORTE */



/* ===== Selection of the UART =====
 * If the AVR supports more than one UART, use 0 for UART0 and 1 for UART1. If
 * you comment out this define, UART0 is used as default.
 */

//#define USE_UART 0

/* ===== Configure WDT Reset handling =====
 * Uncomment the line below too ignore WDT resets. If the BL_PIN is low
 * (active), while a WDT reset occures, the bootload will ignore it and start
 * the application.
 */
//#define HANDLE_WDT_RESET 1


/* ===== Configure QUIT behaviour =====
 * After flashing the application, programmers like AVRDUDE send a 'Quit'
 * command. ATMegaBOOT normally lets the LED flash four times.
 *
 * The default behaviour is, that you have to set the BL_PIN to high and reset
 * the hardware to start the application.
 *
 * There are two different ways to handle this. Uncomment WANT_START_APP if the
 * application should be started right after the quit. Additionally, uncomment
 * WANT_WAIT_BL to wait with the automatic start of the application, until the
 * BL_PIN is high again! So WANT_WAIT_BL alone doesn't make sense.
 *
 */
//#define WANT_WAIT_BL 1
#define WANT_START_APP 1


/* ===== Configuration of the port of the MCU =====
 *
 * To avoid hardware damage while the bootloader is active, it's possible to
 * define the DDR and PORT register of every port of the MCU. Select the
 * port which must have another than the powerup defaults.
 *
 * To enable the setup of an port, uncomment INIT_PORT_*!
 */

#define INIT_PORT_A
#define DIR_PORT_A  BIN2BYTE(00011110)
#define VAL_PORT_A  BIN2BYTE(01111100)

#define INIT_PORT_B
#define DIR_PORT_B  BIN2BYTE(11111110)
#define VAL_PORT_B  BIN2BYTE(00011101)

#define INIT_PORT_C
#define DIR_PORT_C  BIN2BYTE(11001111)
#define VAL_PORT_C  BIN2BYTE(00110000)

#define INIT_PORT_D
#define DIR_PORT_D  BIN2BYTE(11111011)
#define VAL_PORT_D  BIN2BYTE(10001100)

/* #define INIT_PORT_E */
/* #define DIR_PORT_E  BIN2BYTE(00000000) */
/* #define VAL_PORT_E  BIN2BYTE(11111111) */

/* #define INIT_PORT_F */
/* #define DIR_PORT_F  BIN2BYTE(00000000) */
/* #define VAL_PORT_F  BIN2BYTE(11111111) */

/* #define INIT_PORT_G */
/* #define DIR_PORT_G  BIN2BYTE(00000000) */
/* #define VAL_PORT_G  BIN2BYTE(11111111) */

/* #define INIT_PORT_H */
/* #define DIR_PORT_H  BIN2BYTE(00000000) */
/* #define VAL_PORT_H  BIN2BYTE(11111111) */



/* ==[End of file]========================================================== */


projects/config_RFIDLOCK.h
/* config_RFIDLOCK.h -- product specific bootloader configuration header.
 *
 * This header contains declarations and defines for a special product. To
 * enable the usage of this header, uncomment USE_CONFIG_H in the
 * product_TEMPLATE.mak!
 */


/* -------------------------------------------------------------------------
 * Some macros used by this header..
 */

/* This macro converts a sequence of 0 and 1 into a BYTE. The sequence start
 * with the MSB! Make shure that you allways define all 8 bits!
 * Sample: BIN2BYTE(11001010) is converted to 0xCA (1100->C & 1010->A).
 */
#define BIN2BYTE(a) ( ((0x##a##L>>21) & 0x80) + ((0x##a##L>>18) & 0x40) \
                    + ((0x##a##L>>15) & 0x20) + ((0x##a##L>>12) & 0x10) \
                    + ((0x##a##L>>9) & 0x08) + ((0x##a##L>>6) & 0x04)  \
                    + ((0x##a##L>>3) & 0x02) + (0x##a##L & 0x01))



/* -------------------------------------------------------------------------
 * CONFIGURATION
 */



/* ===== Configure Monitor Welcome =====
 * Uncomment to define the welcome message of the Monitor.
 */
/* #define WELCOME_MSG "MegaBOOT Monitor\r\n" */



/* ===== Configure the RS485 stuff =====
 * If your UART is a RS485, define the port-, the DDR-register and the pin number
 * of the output which enables (with 1) the RS485 transmitter. If you don't use a
 * RS485, comment out these three lines.
 */
/* #define RS485_TXON  PINE2 */
/* #define RS485_DDR   DDRE */
/* #define RS485_PORT  PORTE */



/* ===== Selection of the UART =====
 * If the AVR supports more than one UART, use 0 for UART0 and 1 for UART1. If
 * you comment out this define, UART0 is used as default.
 */
#define USE_UART 1

/* ===== Configure WDT Reset handling =====
 * Uncomment the line below too ignore WDT resets. If the BL_PIN is low
 * (active), while a WDT reset occures, the bootload will ignore it and start
 * the application.
 */
//#define HANDLE_WDT_RESET 1


/* ===== Configure QUIT behaviour =====
 * After flashing the application, programmers like AVRDUDE send a 'Quit'
 * command. ATMegaBOOT normally lets the LED flash four times.
 *
 * The default behaviour is, that you have to set the BL_PIN to high and reset
 * the hardware to start the application.
 *
 * There are two different ways to handle this. Uncomment WANT_START_APP if the
 * application should be started right after the quit. Additionally, uncomment
 * WANT_WAIT_BL to wait with the automatic start of the application, until the
 * BL_PIN is high again! So WANT_WAIT_BL alone doesn't make sense.
 *
 */
//#define WANT_WAIT_BL 1
#define WANT_START_APP 1


/* ===== Configuration of the port of the MCU =====
 *
 * To avoid hardware damage while the bootloader is active, it's possible to
 * define the DDR and PORT register of every port of the MCU. Select the
 * port which must have another than the powerup defaults.
 *
 * To enable the setup of an port, uncomment INIT_PORT_*!
 */

#define INIT_PORT_A
#define DIR_PORT_A  BIN2BYTE(00011110)
#define VAL_PORT_A  BIN2BYTE(01100000)

#define INIT_PORT_B
#define DIR_PORT_B  BIN2BYTE(11111110)
#define VAL_PORT_B  BIN2BYTE(00011001)

#define INIT_PORT_C
#define DIR_PORT_C  BIN2BYTE(11001111)
#define VAL_PORT_C  BIN2BYTE(00110000)

#define INIT_PORT_D
#define DIR_PORT_D  BIN2BYTE(11111011)
#define VAL_PORT_D  BIN2BYTE(10001100)

/* #define INIT_PORT_E */
/* #define DIR_PORT_E  BIN2BYTE(00000000) */
/* #define VAL_PORT_E  BIN2BYTE(11111111) */

/* #define INIT_PORT_F */
/* #define DIR_PORT_F  BIN2BYTE(00000000) */
/* #define VAL_PORT_F  BIN2BYTE(11111111) */

/* #define INIT_PORT_G */
/* #define DIR_PORT_G  BIN2BYTE(00000000) */
/* #define VAL_PORT_G  BIN2BYTE(11111111) */

/* #define INIT_PORT_H */
/* #define DIR_PORT_H  BIN2BYTE(00000000) */
/* #define VAL_PORT_H  BIN2BYTE(11111111) */



/* ==[End of file]========================================================== */


projects/config_SAVIMASTER.h
/* config_TEMPLATE.h -- product specific bootloader configuration header.
 *
 * This header contains declarations and defines for a special product. To
 * enable the usage of this header, uncomment USE_CONFIG_H in the
 * product_TEMPLATE.mak!
 */


/* -------------------------------------------------------------------------
 * Some macros used by this header..
 */

/* This macro converts a sequence of 0 and 1 into a BYTE. The sequence start
 * with the MSB! Make shure that you allways define all 8 bits!
 * Sample: BIN2BYTE(11001010) is converted to 0xCA (1100->C & 1010->A).
 */
#define BIN2BYTE(a) ( ((0x##a##L>>21) & 0x80) + ((0x##a##L>>18) & 0x40) \
                    + ((0x##a##L>>15) & 0x20) + ((0x##a##L>>12) & 0x10) \
                    + ((0x##a##L>>9) & 0x08) + ((0x##a##L>>6) & 0x04)  \
                    + ((0x##a##L>>3) & 0x02) + (0x##a##L & 0x01))



/* -------------------------------------------------------------------------
 * CONFIGURATION
 */



/* ===== Configure Monitor Welcome =====
 * Uncomment to define the welcome message of the Monitor.
 */
#define WELCOME_MSG "SaVi.MASTER Monitor\r\n"



/* ===== Configure the RS485 stuff =====
 * If your UART is a RS485, define the port-, the DDR-register and the pin number
 * of the output which enables (with 1) the RS485 transmitter. If you don't use a
 * RS485, comment out these three lines.
 */
#define RS485_TXON  PINE2
#define RS485_DDR   DDRE
#define RS485_PORT  PORTE



/* ===== Selection of the UART =====
 * If the AVR supports more than one UART, use 0 for UART0 and 1 for UART1. If
 * you comment out this define, UART0 is used as default.
 */
#define USE_UART 0

/* ===== Configure WDT Reset handling =====
 * Uncomment the line below too ignore WDT resets. If the BL_PIN is low
 * (active), while a WDT reset occures, the bootload will ignore it and start
 * the application.
 */
//#define HANDLE_WDT_RESET 1


/* ===== Configure QUIT behaviour =====
 * After flashing the application, programmers like AVRDUDE send a 'Quit'
 * command. ATMegaBOOT normally lets the LED flash four times.
 *
 * The default behaviour is, that you have to set the BL_PIN to high and reset
 * the hardware to start the application.
 *
 * There are two different ways to handle this. Uncomment WANT_START_APP if the
 * application should be started right after the quit. Additionally, uncomment
 * WANT_WAIT_BL to wait with the automatic start of the application, until the
 * BL_PIN is high again! So WANT_WAIT_BL alone doesn't make sense.
 *
 */
//#define WANT_WAIT_BL 1
#define WANT_START_APP 1


/* ===== Configuration of the port of the MCU =====
 *
 * To avoid hardware damage while the bootloader is active, it's possible to
 * define the DDR and PORT register of every port of the MCU. Select the
 * port which must have another than the powerup defaults.
 *
 * To enable the setup of an port, uncomment INIT_PORT_*!
 */
#define INIT_PORT_A
#define DIR_PORT_A  BIN2BYTE(11111111)
#define VAL_PORT_A  BIN2BYTE(00000000)
#define INIT_PORT_B
#define DIR_PORT_B  BIN2BYTE(11111111)
#define VAL_PORT_B  BIN2BYTE(00001101)
#define INIT_PORT_C
#define DIR_PORT_C  BIN2BYTE(00000000)
#define VAL_PORT_C  BIN2BYTE(11111111)
#define INIT_PORT_D
#define DIR_PORT_D  BIN2BYTE(00010110)
#define VAL_PORT_D  BIN2BYTE(11101111)
#define INIT_PORT_E
#define DIR_PORT_E  BIN2BYTE(00011110)
#define VAL_PORT_E  BIN2BYTE(11100011)
#define INIT_PORT_F
#define DIR_PORT_F  BIN2BYTE(00000000)
#define VAL_PORT_F  BIN2BYTE(00000001)
#define INIT_PORT_G
#define DIR_PORT_G  BIN2BYTE(00001111)
#define VAL_PORT_G  BIN2BYTE(00000000)



/* ==[End of file]========================================================== */


config_TEMPLATE.h
/* config_TEMPLATE.h -- product specific bootloader configuration header.
 *
 * This header contains declarations and defines for a special product. To
 * enable the usage of this header, uncomment USE_CONFIG_H in the
 * product_TEMPLATE.mak!
 */


/* -------------------------------------------------------------------------
 * Some macros used by this header..
 */

/* This macro converts a sequence of 0 and 1 into a BYTE. The sequence start
 * with the MSB! Make shure that you allways define all 8 bits!
 * Sample: BIN2BYTE(11001010) is converted to 0xCA (1100->C & 1010->A).
 */
#define BIN2BYTE(a) ( ((0x##a##L>>21) & 0x80) + ((0x##a##L>>18) & 0x40) \
                    + ((0x##a##L>>15) & 0x20) + ((0x##a##L>>12) & 0x10) \
                    + ((0x##a##L>>9) & 0x08) + ((0x##a##L>>6) & 0x04)  \
                    + ((0x##a##L>>3) & 0x02) + (0x##a##L & 0x01))



/* -------------------------------------------------------------------------
 * CONFIGURATION
 */



/* ===== Configure Monitor Welcome =====
 * Uncomment to define the welcome message of the Monitor.
 */
/* #define WELCOME_MSG "MegaBOOT Monitor\r\n" */


/* ===== Configure reading signature =====
 * Uncomment this to enable the UNIVERSAL command of the STK500 protocol
 * to return the signature. This method is used by AVRDUDE.
 */
#define WANT_AVRDUDE_SIGNATURE 1


/* ===== Configure the RS485 stuff =====
 * If your UART is a RS485, define the port-, the DDR-register and the pin number
 * of the output which enables (with 1) the RS485 transmitter. If you don't use a
 * RS485, comment out these three lines.
 */
/* #define RS485_TXON  PINE2 */
/* #define RS485_DDR   DDRE */
/* #define RS485_PORT  PORTE */


/* ===== Selection of the UART =====
 * If the AVR supports more than one UART, use 0 for UART0 and 1 for UART1. If
 * you comment out this define, UART0 is used as default.
 */
#define USE_UART 0


/* ===== Configure "forced enter" mode =====
 * In the case there is no bootloader pin, uncommenting this flag configures
 * the bootloader to use "forced enter". Therefore the booloader waits up to
 * 3 seconds for a sequence of 5+ consecutive '*' characters at the UART.
 * If this '*' sequence is received, the bootloader starts sending '*' and waits
 * for an empty UART buffer. If no more data is fetched, the regular bootloader
 * is entered.
 */
//#define USE_FORCED_BOOTLOAD_ENTER 1


/* ===== Configure WDT Reset handling =====
 * Uncomment the line below too ignore WDT resets. If the BL_PIN is low
 * (active), while a WDT reset occures, the bootload will ignore it and start
 * the application.
 */
//#define HANDLE_WDT_RESET 1


/* ===== Configure QUIT behaviour =====
 * After flashing the application, programmers like AVRDUDE send a 'Quit'
 * command. ATMegaBOOT normally lets the LED flash four times.
 *
 * The default behaviour is, that you have to set the BL_PIN to high and reset
 * the hardware to start the application.
 *
 * There are two different ways to handle this. Uncomment WANT_START_APP if the
 * application should be started right after the quit. Additionally, uncomment
 * WANT_WAIT_BL to wait with the automatic start of the application, until the
 * BL_PIN is high again! So WANT_WAIT_BL alone doesn't make sense.
 *
 */
//#define WANT_WAIT_BL 1
//#define WANT_START_APP 1


/* ===== Configuration of the port of the MCU =====
 *
 * To avoid hardware damage while the bootloader is active, it's possible to
 * define the DDR and PORT register of every port of the MCU. Select the
 * port which must have another than the powerup defaults.
 *
 * To enable the setup of an port, uncomment INIT_PORT_*!
 */

/* #define INIT_PORT_A */
/* #define DIR_PORT_A  BIN2BYTE(00000000) */
/* #define VAL_PORT_A  BIN2BYTE(11111111) */

/* #define INIT_PORT_B */
/* #define DIR_PORT_B  BIN2BYTE(00000000) */
/* #define VAL_PORT_B  BIN2BYTE(11111111) */

/* #define INIT_PORT_C */
/* #define DIR_PORT_C  BIN2BYTE(00000000) */
/* #define VAL_PORT_C  BIN2BYTE(11111111) */

/* #define INIT_PORT_D */
/* #define DIR_PORT_D  BIN2BYTE(00000000) */
/* #define VAL_PORT_D  BIN2BYTE(11111111) */

/* #define INIT_PORT_E */
/* #define DIR_PORT_E  BIN2BYTE(00000000) */
/* #define VAL_PORT_E  BIN2BYTE(11111111) */

/* #define INIT_PORT_F */
/* #define DIR_PORT_F  BIN2BYTE(00000000) */
/* #define VAL_PORT_F  BIN2BYTE(11111111) */

/* #define INIT_PORT_G */
/* #define DIR_PORT_G  BIN2BYTE(00000000) */
/* #define VAL_PORT_G  BIN2BYTE(11111111) */

/* #define INIT_PORT_H */
/* #define DIR_PORT_H  BIN2BYTE(00000000) */
/* #define VAL_PORT_H  BIN2BYTE(11111111) */



/* ==[End of file]========================================================== */
